src/vm/gccover.cpp

```diff
@@ -250,6 +341,88 @@ void SetupGcCoverageForNativeImage(Module* module)
 }
 #endif
 
+
+void ReplaceInstrAfterCall(SLOT instrToReplace, MethodDesc* callMD)
+{
+    ReturnKind returnKind = callMD->GetReturnKind(true);
+    if (!IsValidReturnKind(returnKind))
+    {
+#if !defined(_TARGET_AMD64_) || !defined(PLATFORM_UNIX)
+        return;
+#else // _TARGET_AMD64_ && PLATFORM_UNIX
+        _ASSERTE("Unexpected return kind for x64 Unix.");
+#endif // _TARGET_AMD64_ && PLATFORM_UNIX
+    }
+    _ASSERTE(IsValidReturnKind(returnKind));
+
+    bool pointerKind = IsPointerReturnKind(returnKind);
+#ifdef _TARGET_ARM_
+    size_t instrLen = GetARMInstructionLength(instrToReplace);
+    bool protectReturn = pointerKind;
+    if (protectReturn)
+        if (instrLen == 2)
+            *(WORD*)instrToReplace = INTERRUPT_INSTR_PROTECT_RET;
+        else
+            *(DWORD*)instrToReplace = INTERRUPT_INSTR_PROTECT_RET_32;
+    else
+        if (instrLen == 2)
+            *(WORD*)instrToReplace = INTERRUPT_INSTR;
+        else
+            *(DWORD*)instrToReplace = INTERRUPT_INSTR_32;
+#elif defined(_TARGET_ARM64_)
+    bool protectReturn = pointerKind;
+    if (protectReturn)
+        *(DWORD*)instrToReplace = INTERRUPT_INSTR_PROTECT_RET;
+    else
+        *(DWORD*)instrToReplace = INTERRUPT_INSTR;
+#elif defined(_TARGET_AMD64_) || defined(_TARGET_X86_)
+
+
+    if (pointerKind)
+    {
+        bool protectRegister[2] = { false };
+
+        int regNo = 0;
+        bool moreRegisters = false;
+        do
+        {
+            ReturnKind fieldKind = ExtractRegReturnKind(returnKind, regNo, moreRegisters);
+            if (IsPointerFieldReturnKind(fieldKind))
+            {
+                protectRegister[regNo] = true;
+            }
+            regNo++;
+        } while (moreRegisters);
+
+        if (protectRegister[0] && !protectRegister[1])
+        {
+            *instrToReplace = INTERRUPT_INSTR_PROTECT_FIRST_RET;
+        }
+        else
+        {
+#if !defined(_TARGET_AMD64_) || !defined(PLATFORM_UNIX)
+            _ASSERTE(!"Not expected multi reg return with pointers.");
+#endif // !_TARGET_AMD64_ || !PLATFORM_UNIX
+            if (!protectRegister[0] && protectRegister[1])
+            {
+                *instrToReplace = INTERRUPT_INSTR_PROTECT_SECOND_RET;
+            }
+            else
+            {
+                _ASSERTE(protectRegister[0] && protectRegister[1]);
+                *instrToReplace = INTERRUPT_INSTR_PROTECT_BOTH_RET;
```

Comment:
I would stay it as is, because we need to check that `protectRegister[1] == false` on `!_TARGET_AMD64_ || !PLATFORM_UNIX` and we can't have `!protectRegister[0] && ~protectRegister[1]` case here.